### PR TITLE
Migrate quicklist unit test to new framework

### LINKED ISSUE #428: Migrate unit tests to new framework
The legacy unit tests were a collection of distinct tests stored at the bottom of various files, that required rebuilding the server with a special flag and starting the server. We're moving to a new [framework](https://github.com/valkey-io/valkey/pull/344) that can always be built and are running on a unified infrastructure. This issue is keeping track of migration from the old to new framework.

To migrate the tests, remove the custom testing code in the testing file lists listed below, create a new test like src/unit/test_*.c, and remove the reference in https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/server.c#L6890. See the [README](https://github.com/valkey-io/valkey/blob/unstable/src/unit/README.md) for more information about writing a new unit test. 

- [x] : [Zipmap.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/zipmap.c#L444) https://github.com/valkey-io/valkey/pull/474
- [X] : [util.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/util.c#L1378) https://github.com/valkey-io/valkey/pull/448
- [x] : [kvstore.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/kvstore.c#L869) https://github.com/valkey-io/valkey/pull/446
- [x] : [dict.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/dict.c#L1758)https://github.com/valkey-io/valkey/pull/673 and https://github.com/valkey-io/valkey/pull/946
- [X] : [endianconv.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/endianconv.c#L104) https://github.com/valkey-io/valkey/pull/458
- [x] : [quicklist.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/quicklist.c#L215) https://github.com/valkey-io/valkey/pull/515
- [X] : [sds.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/sds.c#L1232) https://github.com/valkey-io/valkey/pull/478
- [x] : [sha1.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/sha1.c#L211) (https://github.com/valkey-io/valkey/pull/470)
- [x] : [ziplist.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/ziplist.c#L1689) https://github.com/valkey-io/valkey/pull/486
- [x] : [zmalloc.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/zmalloc.c#L911) https://github.com/valkey-io/valkey/pull/493
- [x] : [listpack.c](https://github.com/valkey-io/valkey/blob/5b1fd222ed97430dfeef27a95aeb7e1fb8fd1352/src/listpack.c#L1682) (https://github.com/valkey-io/valkey/pull/484) and #949
- [x] : Cleanup remaining references of SERVER_TEST (#515)

### PR DESCRIPTION
**Pull request by artikell**
Migrate quicklist unit test to new unit test framework, and cleanup
remaining references of SERVER_TEST, parent ticket #428.

Closes #428.

### CODE REVIEW

**Comment by artikell** [comment]
@madolson I need your help to review it. I have broken it down into multiple commitments. Used to assist with code review.
There are several key points involved in this:
- I tried to include quicklist. c
- Used [python script](https://gist.github.com/artikell/e84dd3f92ae18e5ea4da7b24b5c5eed1) to replace function
- Continuing to use err to determine if there is an error

**Comment by artikell** [comment]
@enjoy-binbin @madolson It seems that there are only SERVER_TEST left in quicklist, I want to push this forward. There are more cases in this part. Do we want to merge in batches, or is it more appropriate to merge with one mr?

**Comment by enjoy-binbin** [comment]
I think one PR should be enough, there is no need for more batches, please go ahead, thanks

**Comment by artikell** [comment]
> I think one PR should be enough, there is no need for more batches, please go ahead, thanks

The conflict has been resolved. Could you kindly review it? Thank you

File: src/quicklist.c [hunk 1]
```diff
@@ -209,10 +209,9 @@ void quicklistRelease(quicklist *quicklist) {
 /* Compress the listpack in 'node' and update encoding details.
  * Returns 1 if listpack compressed successfully.
  * Returns 0 if compression failed or if listpack too small to compress. */
+
```
Comment (enjoy-binbin) [inline src/quicklist.c]: plrease remove these extra line.

Comment (artikell) [reply src/quicklist.c]: This has been dealt with.

File: src/unit/test_quicklist.c [hunk 2]
```diff
@@ -0,0 +1,2304 @@
...
+#include "../zmalloc.h"
+#include "../listpack.h"
+#include "../quicklist.c"
+#include "../util.h"
+
+static int options[] = {0, 1, 2, 3, 4, 5, 6, 10};
+static int option_count = 8;
+
+static int fills[] = {-5, -4, -3, -2, -1, 0,
+                      1, 2, 32, 66, 128, 999};
```
Comment (enjoy-binbin) [inline src/unit/test_quicklist.c:16]:
```suggestion
static int fills[] = {-5, -4, -3, -2, -1, 0, 1, 2, 32, 66, 128, 999};
```

**Review by enjoy-binbin — APPROVED** [review]
thanks, i pushed a dummy commit to trigger the CI

**Comment by enjoy-binbin** [comment]
CI report some failure, please take a look:
https://github.com/valkey-io/valkey/actions/runs/11716457128/job/32634549755?pr=515#step:4:560
```
sed '/Everything below this line/,$d' fmtargs.h > fmtargs.h.tmp
/usr/bin/python3 ../utils/generate-fmtargs.py >> fmtargs.h.tmp
mv fmtargs.h.tmp fmtargs.h
    CC server.o
server.c:6799:19: error: ISO C forbids empty initializer braces [-Werror=pedantic]
 6799 | } serverTests[] = {};
      |                   ^
server.c:6799:3: error: zero or negative size array ‘serverTests’
 6799 | } serverTests[] = {};
      |   ^~~~~~~~~~~
```

**Comment by artikell** [comment]
> CI report some failure, please take a look: https://github.com/valkey-io/valkey/actions/runs/11716457128/job/32634549755?pr=515#step:4:560
> 
> ```
> sed '/Everything below this line/,$d' fmtargs.h > fmtargs.h.tmp
> /usr/bin/python3 ../utils/generate-fmtargs.py >> fmtargs.h.tmp
> mv fmtargs.h.tmp fmtargs.h
>     CC server.o
> server.c:6799:19: error: ISO C forbids empty initializer braces [-Werror=pedantic]
>  6799 | } serverTests[] = {};
>       |                   ^
> server.c:6799:3: error: zero or negative size array ‘serverTests’
>  6799 | } serverTests[] = {};
>       |   ^~~~~~~~~~~
> ```

ok，Let me clean this up together

**Comment by enjoy-binbin** [comment]
i see daily.ci is still have some `valkey-server test all`, please also remove it, and also use the new `./src/valkey-unit-tests` to replace it. please also remove all `DSERVER_TEST`

**Comment by artikell** [comment]
> i see daily.ci is still have some `valkey-server test all`, please also remove it, and also use the new `./src/valkey-unit-tests` to replace it. please also remove all `DSERVER_TEST`

@enjoy-binbin Is there a way to retry a process alone? It always seems to get stuck in strange places.

**Review by enjoy-binbin — APPROVED** [review]
thanks for the work, LGTM, i did a minor cleanup in the last commit.

ps. please avoid the force push, it makes it difficult to catch incremental diffs (commits). I will retry the test if it fail

(i force push for the DCO, suck).

**Comment by enjoy-binbin** [comment]
I see the top comment is outdated, could you please update the top comment to reflect the changes? And then i believe we are good to go.

**Pull request merged**